pkg/resource/managed_reconciler.go

```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+	"github.com/crossplaneio/crossplane/pkg/meta"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+const (
+	managedControllerName   = "managedresource.crossplane.io"
+	managedFinalizerName    = "finalizer." + managedControllerName
+	managedReconcileTimeout = 1 * time.Minute
+
+	defaultManagedShortWait = 30 * time.Second
+	defaultManagedLongWait  = 1 * time.Minute
+)
+
+// Error strings.
+const (
+	errGetManaged    = "cannot get managed resource"
+	errPublishSecret = "cannot publish managed resource connection secret"
+)
+
+// ConnectionDetails created or updated during an operation on an external
+// resource, for example usernames, passwords, endpoints, ports, etc.
+type ConnectionDetails map[string][]byte
+
+// A ManagedPublisher publishes the supplied ConnectionDetails for the supplied
+// Managed resource. ManagedPublishers must handle the case in which the
+// supplied ConnectionDetails are empty.
+type ManagedPublisher interface {
```

Comment:
Good point. Connection details are in practice always published as a `Secret` today, but this design tries to abstract that fact away in preparation for a future in which we might publish them in other ways. The abstraction leaks in that we still rely on Kubernetes garbage collection to remove the secrets.

My preference would be to avoid `ExternalClient` implementations communicating with the Kubernetes API directly; I like the separation of concern in which `ExternalClient` implementations are truly clients of the external system, not clients of the external system _and_ the Crossplane API server. Given this, I prefer the `Unpublish` option because it avoids the `ExternalClient` potentially having to connect to the API server to publish connection details.

Another direction that may be worth exploring is to simply require, by convention, that `ManagedPublisher` implementations ensure any artifacts they create will be garbage collected by the Kubernetes API when the managed resource is deleted. This would limit implementations to creating objects, with appropriately controller references, in the API server.

Reply:
`Unpublish` makes more sense to me as well. They'd just do no-op in that function if they handed it over to kubernetes.

Enforcement by convention would be harder to sustain given that we're considering separate repos which will have small overlap of contributors. If we have a way to require it explicitly, I think we should do it.

Reply:
I like the idea of an `Unpublish` interface, that would be a no-op in the case of garbage collected Kubernetes resources. I'm not sure about the names though. Perhaps:

```go
type ManagedConnectionPublisher interface {
    PublishConnection(ctx context.Context, mg Managed, c ConnectionData) error
}

type ManagedConnectionRedacter interface {
   RedactConnection(ctx context.Context, mg Managed, c ConnectionData) 
}
```

Redact is the closest word I can think of that means roughly "unpublish".

Reply:
I think `Redact` would be intuitive if it was expected to delete only some keys from the secret. But as far as I see, `Publish` creates & updates the secret object and `Redact` deletes it.

How about `ApplyConnection` and `DeleteConnection`? Sounds similar to kubernetes verbs and implies that the operation covers object-level actions as well. `ApplyConnection` doesn't sound very natural though.

Another point, I think they should live in the same interface so that we can safely assume that both actions are considered by the developer.

Reply:
Here is a quick pass on it to give an idea how it could look like https://github.com/muvaf/crossplane/commit/531e28e6ca44e2e0bad78096c7ae897f382172f0

Reply:
Yeah, I can see your point about the meaning of `Redact`. I'd prefer to keep `Publish` because it has some symmetry with the terms we've been using in the draft resource usage design in https://github.com/crossplaneio/crossplane/pull/564.

https://en.wiktionary.org/wiki/unpublish has convinced me that `Unpublish` is a real thing, so I'd prefer your original suggestion of `PublishConnection` and `UnpublishConnection`.